### PR TITLE
Remove UbuntuVersions function

### PR DESCRIPTION
**Pull request by jack-w-shaw**
This function is one of the few left that uses distro-info to determine it's output.

This one, however, returns series and bases which both are and are not supported. We wish to replace this whole system with just a hardcoded list of supported bases. As such, this function needs to go

Fortunately, it was only used in 1 place. upgradevalidation. We used this function to get a list of all unsupported ubuntu versions, which we would then use to query state to check if any machines are deployed to those bases.

Instead, we use WorkloadBases to get the supported bases, and then assert that all machines are deployed to one of these bases.

In doing so, add "AllMachinesCount" state method, which returns an integer. This isn't ideal, since we should be trying to avoid making changes to state whilst it's being migrated to DQLite, but it turns out the simpler signature removes a lot more complexity elsewhere. We no longer need to shim out machines in our facades, and transform between slices of identical interfaces, etc.

As a flyby, remove non-Ubuntu support for non-Ubuntu workloads. This was a planned change in 4.0

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing
- [ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

### Happy path
```
$ juju bootstrap lxd lxd
$ juju add-model m
$ juju upgrade-controller --build-agent
no prepackaged agent binaries available, using local agent binary 4.0-beta4.2 (built from source)
best version:
    4.0-beta4.2
started upgrade to 4.0-beta4.2

$ juju upgrade-model
best version:
    4.0-beta4.2
started upgrade to 4.0-beta4.2

$ juju add-machine --base ubuntu@20.04
$ juju add-machine --base ubuntu@22.04
$ juju status
Model  Controller  Cloud/Region         Version      Timestamp
m      lxd         localhost/localhost  4.0-beta4.2  14:52:18+01:00

Machine  State    Address         Inst id        Base          AZ  Message
0        started  10.219.211.191  juju-3aa90a-0  ubuntu@20.04      Running
1        started  10.219.211.116  juju-3aa90a-2  ubuntu@22.04      Running

$ juju upgrade-controller --build-agent
no prepackaged agent binaries available, using local agent binary 4.0-beta4.2 (built from source)
best version:
    4.0-beta4.2
started upgrade to 4.0-beta4.2

$ juju upgrade-model
best version:
    4.0-beta4.2
started upgrade to 4.0-beta4.2
```

### Sad path

```
$ juju add-model m2
$ juju add-machine --base ubuntu@18.04
$ juju status
Model  Controller  Cloud/Region         Version      Timestamp
m2     lxd         localhost/localhost  4.0-beta4.2  14:56:29+01:00

Machine  State    Address         Inst id        Base          AZ  Message
0        started  10.219.211.213  juju-4e641d-1  ubuntu@18.04      Running

$ juju upgrade-controller --build-agent
no prepackaged agent binaries available, using local agent binary 4.0-beta4.2 (built from source)
ERROR cannot upgrade to "4.0-beta4.2" due to issues with these models:
"admin/m2":
- the model hosts 1 ubuntu machine(s) with an unsupported base. The supported bases are: ubuntu@20.04, ubuntu@22.04, ubuntu@24.04

$ juju remove-machine 0
$ juju add-machine --base centos@7
$ juju status
Model  Controller  Cloud/Region         Version      Timestamp
m2     lxd         localhost/localhost  4.0-beta4.2  14:57:51+01:00

Machine  State  Address  Inst id  Base      AZ  Message
2        down            pending  centos@7      no matching agent binaries available

$ juju upgrade-controller --build-agent
no prepackaged agent binaries available, using local agent binary 4.0-beta4.2 (built from source)
ERROR cannot upgrade to "4.0-beta4.2" due to issues with these models:
"admin/m2":
- the model hosts 1 ubuntu machine(s) with an unsupported base. The supported bases are: ubuntu@20.04, ubuntu@22.04, ubuntu@24.04
```

### CODE REVIEW

File: state/state.go [hunk 1]
```diff
@@ -743,6 +743,14 @@ func (st *State) AllMachines() ([]*Machine, error) {
 	return st.allMachines(machinesCollection)
 }
```
Comment (Aflynn50) [inline state/state.go:745]: Can you put a comment on this.

File: core/base/supported.go [hunk 2]
```diff
@@ -398,12 +398,12 @@ var centosSeries = map[SeriesName]seriesVersion{
...
+		Supported:    false,
 	},
 	Centos9: {
 		WorkloadType: OtherWorkloadType,
 		OS:           CentosOS,
 		Version:      "9",
-		Supported:    true,
+		Supported:    false,
 	},
 }
```
Comment (SimonRichardson) [inline core/base/supported.go:409]: If we set this to false, then are we dropping support for centos in 4 completely? If that's the case we should just remove this list.

File: core/base/supportedbases.go [hunk 3]
```diff
@@ -76,26 +76,3 @@ func supportedInfoForType(path string, now time.Time, requestedBase Base, imageS
 
 	return supported, nil
 }
-
-// UbuntuVersions returns the ubuntu versions as a map.
-func UbuntuVersions(supported, esmSupported *bool) map[string]string {
-	return ubuntuVersions(supported, esmSupported, ubuntuSeries)
-}
```
Comment (SimonRichardson) [inline core/base/supportedbases.go:83]: Glad to see this go. Once we have a hardcoded list, we don't need to ask these questions.

File: internal/upgrades/upgradevalidation/validation.go [hunk 4]
```diff
@@ -152,41 +153,39 @@ func getCheckUpgradeSeriesLockForModel(force bool) Validator {
...
-		keys = append(keys, k)
-	}
-	sort.Strings(keys)
-	var output []string
-	for _, k := range keys {
-		output = append(output, fmt.Sprintf("%s(%d)", k, result[k]))
-	}
-	return strings.Join(output, " ")
-}
+// For testing.
```
Comment (Aflynn50) [inline internal/upgrades/upgradevalidation/validation.go:156]: I guess this shouldn't be here?

File: core/base/supportedseries.go [hunk 5]
```diff
@@ -82,12 +82,6 @@ func composeSeriesVersions() {
 	for k, v := range centosSeries {
 		allSeriesVersions[k] = v
 	}
-	allSeriesVersions[genericLinuxSeries] = seriesVersion{
-		WorkloadType: OtherWorkloadType,
-		OS:           genericLinuxOS,
-		Version:      genericLinuxVersion,
-		Supported:    true,
-	}
```
Comment (SimonRichardson) [inline core/base/supportedseries.go:90]: Removal of generic linux means we don't support anything outside of ubuntu. Is this correct?

File: internal/upgrades/upgradevalidation/validation.go [hunk 6]
```diff
@@ -152,41 +153,39 @@ func getCheckUpgradeSeriesLockForModel(force bool) Validator {
...
+	for _, v := range baseCountMap {
+		allSupportedCount += v
+	}
+
+	allMachinesCount, err := st.AllMachinesCount()
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot get all machines count")
+	}
+
+	if totalUnsupported := allMachinesCount - allSupportedCount; totalUnsupported > 0 {
```
Comment (Aflynn50) [inline internal/upgrades/upgradevalidation/validation.go:185]: Noice

File: internal/upgrades/upgradevalidation/validation_test.go [hunk 7]
```diff
@@ -114,12 +117,17 @@ func (s *upgradeValidationSuite) TestCheckForDeprecatedUbuntuSeriesForModel(c *g
...
+
 	st := mocks.NewMockState(ctrl)
-	st.EXPECT().MachineCountForBase(makeBases("ubuntu", ubuntuVersions)).Return(map[string]int{"xenial": 1, "vivid": 2, "trusty": 3}, nil)
+	st.EXPECT().MachineCountForBase(makeBases("ubuntu", []string{"24.04/stable", "22.04/stable", "20.04/stable"})).Return(map[string]int{"ubuntu@20.04": 1, "ubuntu@22.04": 1, "ubuntu@24.04": 2}, nil)
+	st.EXPECT().AllMachinesCount().Return(5, nil)
 
 	blocker, err := upgradevalidation.CheckForDeprecatedUbuntuSeriesForModel("", nil, st, nil)
 	c.Assert(err, jc.ErrorIsNil)
-	c.Assert(blocker.Error(), gc.Equals, `the model hosts deprecated ubuntu machine(s): trusty(3) vivid(2) xenial(1)`)
+	c.Assert(blocker.Error(), gc.Equals, `the model hosts 1 ubuntu machine(s) with an unsupported base. The supported bases are: ubuntu@24.04, ubuntu@22.04, ubuntu@20.04`)
```
Comment (Aflynn50) [inline internal/upgrades/upgradevalidation/validation_test.go:130]: Is there a unit test for the happy path for this?

**Review by Aflynn50 — APPROVED** [review]
Looks good to me, its nice that with this way of doing it juju won't have to have knowledge of unsupported versions.

**Review by SimonRichardson — DISMISSED** [review]
Blocking this until we discuss supported bases.

File: internal/upgrades/upgradevalidation/validation_test.go (hunk 7 above)
Comment (jack-w-shaw) [reply internal/upgrades/upgradevalidation/validation_test.go:130]: Yep, there's a few in that file

**Review by manadart — COMMENTED** [review]
Looks OK, and QA passes, but there are some jobs that need to be made to pass.

File: internal/provider/oci/images.go [hunk 8]
```diff
@@ -258,23 +257,13 @@ func getImageType(img ociCore.Image) ImageType {
...
-		base = corebase.MakeDefaultBase("centos", *img.OperatingSystemVersion)
-		// For the moment, only x86 shapes are supported
-		arch = corearch.AMD64
-	case ubuntuOS:
-		base, arch, isMinimal = parseUbuntuImage(img)
-	default:
+	if osName := *img.OperatingSystem; osName != ubuntuOS {
 		return InstanceImage{}, "", errors.NotSupportedf("os %s", osName)
 	}
+	var (
```
[the remark this answers is not in the document]
Comment (jack-w-shaw) [reply internal/provider/oci/images.go]: Yeah could do, no reason why not. The reason I didn't was to keep the change minimally invasive.

This change was from another PR, which unfortunately has already been merged. I'll slip this in as a flyby here though

**Pull request merged**